File: project/drivers/driver/dart/test/auth_bootstrap_contract_test.dart
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:scratchbird/scratchbird.dart';
import 'package:scratchbird/src/protocol.dart';
import 'package:test/test.dart';

const int _managerProtocolMagic = 0x42444253;
const int _managerProtocolVersion = 0x0101;
const int _managerHeaderSize = 12;
const int _mcpMsgConnectResponse = 0x02;
const int _mcpMsgStatusResponse = 0x64;
const int _mcpMsgHello = 0x65;

void main() {
  ScratchBirdConfig _baseConfig(int port) {
    return ScratchBirdConfig(
      host: '127.0.0.1',
      port: port,
      database: 'db',
      user: 'user',
      password: 'pass',
      sslmode: 'disable',
    );
  }

  test('probeAuthSurface reports direct SCRAM_SHA_512', () async {
    final server = await _LoopbackServer.start((socket, reader) async {
      final startup = await _readProtocolMessage(reader);
      expect(startup.header.type, MessageType.startup);
      await _writeProtocolMessage(
        socket,
        MessageType.authRequest,
        Uint8List.fromList([authScramSha512Method, 0, 0, 0]),
      );
    });

    final probe =
        await ScratchBirdClient.probeAuthSurface(_baseConfig(server.port));
    expect(probe.reachable, isTrue);
    expect(probe.ingressMode, 'direct');
    expect(probe.requiredMethod, 'SCRAM_SHA_512');
    expect(probe.requiredPluginMethodId, 'scratchbird.auth.scram_sha_512');
    expect(probe.admittedMethods.single.executableLocally, isTrue);
    expect(probe.additionalContinuationPossible, isTrue);
    await server.done;
  });

  test('probeAuthSurface reports manager TOKEN ingress', () async {
    final server = await _LoopbackServer.start((socket, reader) async {
      final frame = await _readManagerFrame(reader);
      expect(frame.type, _mcpMsgHello);
      await _writeManagerFrame(
        socket,
        _mcpMsgStatusResponse,
        Uint8List(0),
      );
    });

    final probe = await ScratchBirdClient.probeAuthSurface(
      _baseConfig(server.port).copyWith(frontDoorMode: 'manager_proxy'),
    );
    expect(probe.reachable, isTrue);
    expect(probe.ingressMode, 'manager_proxy');
    expect(probe.requiredMethod, 'TOKEN');
    expect(probe.requiredPluginMethodId, 'scratchbird.auth.authkey_token');
    expect(probe.admittedMethods.single.executableLocally, isTrue);
    await server.done;
  });

  test('connect tracks resolved SCRAM_SHA_512 auth context', () async {
    final server = await _LoopbackServer.start((socket, reader) async {
      final startup = await _readProtocolMessage(reader);
      expect(startup.header.type, MessageType.startup);
      await _writeProtocolMessage(
        socket,
        MessageType.authRequest,
        Uint8List.fromList([authScramSha512Method, 0, 0, 0]),
      );

      final clientFirstMsg = await _readProtocolMessage(reader);
      final clientFirst = utf8.decode(clientFirstMsg.payload);
      final clientNonce = _extractNonce(clientFirst);
      final salt = Uint8List.fromList([1, 2, 3, 4, 5, 6, 7, 8]);
      final serverFirst =
          'r=${clientNonce}server,s=${base64.encode(salt)},i=4096';
      final continuePayload = BytesBuilder()
        ..add([authScramSha512Method, 1, 0, 0])
        ..add(_u32(serverFirst.length))
        ..add(utf8.encode(serverFirst));
      await _writeProtocolMessage(
        socket,
        MessageType.authContinue,
        continuePayload.toBytes(),
      );

      final clientFinalMsg = await _readProtocolMessage(reader);
      final clientFinal = utf8.decode(clientFinalMsg.payload);
      final serverFinal = _buildScramServerFinalSha512(
        password: 'pass',
        clientFirst: clientFirst,
        serverFirst: serverFirst,
        clientFinal: clientFinal,
        salt: salt,
        iterations: 4096,
      );
      await _writeProtocolMessage(
        socket,
        MessageType.authOk,
        _makeAuthOkPayload(serverFinal),
        attachmentId: _sessionId(),
      );
      await _writeProtocolMessage(
        socket,
        MessageType.ready,
        Uint8List(0),
        attachmentId: _sessionId(),
      );
    });

    final client = await ScratchBirdClient.connect(_baseConfig(server.port));
    final context = client.getResolvedAuthContext();
    expect(context.ingressMode, 'direct');
    expect(context.resolvedAuthMethod, 'SCRAM_SHA_512');
    expect(context.resolvedAuthPluginId, 'scratchbird.auth.scram_sha_512');
    expect(context.managerAuthenticated, isFalse);
    expect(context.attached, isTrue);
    await client.close();
    await server.done;
  });

  test('connect tracks resolved TOKEN auth context', () async {
    final server = await _LoopbackServer.start((socket, reader) async {
      final startup = await _readProtocolMessage(reader);
      expect(startup.header.type, MessageType.startup);
      await _writeProtocolMessage(
        socket,
        MessageType.authRequest,
        Uint8List.fromList([authTokenMethod, 0, 0, 0]),
      );
      final authResponse = await _readProtocolMessage(reader);
      expect(utf8.decode(authResponse.payload), 'token-123');
      await _writeProtocolMessage(
        socket,
        MessageType.authOk,
        _makeAuthOkPayload(''),
        attachmentId: _sessionId(),
      );
      await _writeProtocolMessage(
        socket,
        MessageType.ready,
        Uint8List(0),
        attachmentId: _sessionId(),
      );
    });

    final client = await ScratchBirdClient.connect(
      _baseConfig(server.port).copyWith(authToken: 'token-123'),
    );
    final context = client.getResolvedAuthContext();
    expect(context.resolvedAuthMethod, 'TOKEN');
    expect(context.resolvedAuthPluginId, 'scratchbird.auth.authkey_token');
    expect(context.attached, isTrue);
    await client.close();
    await server.done;
  });

  test('connect fails closed for PEER auth', () async {
    final server = await _LoopbackServer.start((socket, reader) async {
      final startup = await _readProtocolMessage(reader);
      expect(startup.header.type, MessageType.startup);
      await _writeProtocolMessage(
        socket,
        MessageType.authRequest,
        Uint8List.fromList([authPeerMethod, 0, 0, 0]),
      );
    });

    await expectLater(
      ScratchBirdClient.connect(_baseConfig(server.port)),
      throwsA(
        isA<ScratchBirdAuthException>().having(
          (e) => e.message,
          'message',
          contains('requires external broker support'),
        ),
      ),
    );
    await server.done;
  });
}

class _LoopbackServer {
  final ServerSocket server;
  final Future<void> done;

  _LoopbackServer(this.server, this.done);

  int get port => server.port;

  static Future<_LoopbackServer> start(
    Future<void> Function(Socket socket, _TestSocketReader reader) handler,
  ) async {
    final server = await ServerSocket.bind(InternetAddress.loopbackIPv4, 0);
    final completer = Completer<void>();
    server.listen((socket) async {
      final reader = _TestSocketReader(StreamIterator(socket));
      try {
        await handler(socket, reader);
        await socket.flush();
      } catch (e, st) {
        if (!completer.isCompleted) {
          completer.completeError(e, st);
        }
      } finally {
        await socket.close();
        await server.close();
        if (!completer.isCompleted) {
          completer.complete();
        }
      }
    }, onError: (Object error, StackTrace stackTrace) {
      if (!completer.isCompleted) {
        completer.completeError(error, stackTrace);
      }
    });
    return _LoopbackServer(server, completer.future);
  }
}

class _TestSocketReader {
  final StreamIterator<Uint8List> _iterator;
  final List<int> _buffer = <int>[];

  _TestSocketReader(this._iterator);

  Future<Uint8List> readExact(int length) async {
    while (_buffer.length < length) {
      if (!await _iterator.moveNext()) {
        throw StateError('socket closed before reading $length bytes');
      }
      _buffer.addAll(_iterator.current);
    }
    final bytes = Uint8List.fromList(_buffer.sublist(0, length));
    _buffer.removeRange(0, length);
    return bytes;
  }
}

Future<ScratchBirdMessage> _readProtocolMessage(
    _TestSocketReader reader) async {
  final headerBytes = await reader.readExact(headerSize);
  final header = decodeHeader(headerBytes);
  final payload =
      header.length == 0 ? Uint8List(0) : await reader.readExact(header.length);
  return ScratchBirdMessage(header, payload);
}

Future<void> _writeProtocolMessage(
  Socket socket,
  int type,
  Uint8List payload, {
  Uint8List? attachmentId,
  int txnId = 0,
  int sequence = 0,
}) async {
  socket.add(
    encodeMessage(
      MessageHeader(
        type: type,
        flags: 0,
        length: payload.length,
        sequence: sequence,
        attachmentId: attachmentId ?? Uint8List(16),
        txnId: txnId,
      ),
      payload,
    ),
  );
  await socket.flush();
}

Future<({int type, Uint8List payload})> _readManagerFrame(
  _TestSocketReader reader,
) async {
  final header = await reader.readExact(_managerHeaderSize);
  final data = ByteData.sublistView(header);
  expect(data.getUint32(0, Endian.little), _managerProtocolMagic);
  expect(data.getUint16(4, Endian.little), _managerProtocolVersion);
  final type = data.getUint8(6);
  final payloadLength = data.getUint32(8, Endian.little);
  final payload =
      payloadLength == 0 ? Uint8List(0) : await reader.readExact(payloadLength);
  return (type: type, payload: payload);
}

Future<void> _writeManagerFrame(
    Socket socket, int type, Uint8List payload) async {
  final header = ByteData(_managerHeaderSize);
  header.setUint32(0, _managerProtocolMagic, Endian.little);
  header.setUint16(4, _managerProtocolVersion, Endian.little);
  header.setUint8(6, type);
  header.setUint8(7, 0);
  header.setUint32(8, payload.length, Endian.little);
  socket.add(header.buffer.asUint8List());
  socket.add(payload);
  await socket.flush();
}

String _extractNonce(String clientFirst) {
  for (final part in clientFirst.split(',')) {
    if (part.startsWith('r=')) {
      return part.substring(2);
    }
  }
  throw StateError('missing SCRAM nonce');
}

Uint8List _sessionId() {
  return Uint8List.fromList(List<int>.generate(16, (index) => index + 1));
}

Uint8List _makeAuthOkPayload(String serverInfo) {
  final infoBytes = Uint8List.fromList(utf8.encode(serverInfo));
  final payload = BytesBuilder()
    ..add(_sessionId())
    ..add(_u32(infoBytes.length))
    ..add(infoBytes);
  return payload.toBytes();
}

String _buildScramServerFinalSha512({
  required String password,
  required String clientFirst,
  required String serverFirst,
  required String clientFinal,
  required Uint8List salt,
  required int iterations,
}) {
  final clientFirstBare = clientFirst.substring(3);
  final clientFinalWithoutProof = clientFinal.split(',p=').first;
  final authMessage = '$clientFirstBare,$serverFirst,$clientFinalWithoutProof';
  final salted = _pbkdf2Sha512(utf8.encode(password), salt, iterations, 64);
  final serverKey = _hmacSha512(salted, utf8.encode('Server Key'));
  final signature = _hmacSha512(serverKey, utf8.encode(authMessage));
  return 'v=${base64.encode(signature)}';
}

Uint8List _pbkdf2Sha512(
  List<int> password,
  List<int> salt,
  int iterations,
  int keyLen,
) {
  final blocks = (keyLen / 64).ceil();
  final out = BytesBuilder();
  for (var i = 1; i <= blocks; i++) {
    out.add(_pbkdf2BlockSha512(password, salt, iterations, i));
  }
  final bytes = out.toBytes();
  return Uint8List.sublistView(bytes, 0, keyLen);
}

List<int> _pbkdf2BlockSha512(
  List<int> password,
  List<int> salt,
  int iterations,
  int blockIndex,
) {
  final block = Uint8List(salt.length + 4);
  block.setAll(0, salt);
  block[block.length - 4] = (blockIndex >> 24) & 0xff;
  block[block.length - 3] = (blockIndex >> 16) & 0xff;
  block[block.length - 2] = (blockIndex >> 8) & 0xff;
  block[block.length - 1] = blockIndex & 0xff;
  var u = _hmacSha512(password, block);
  final out = List<int>.from(u);
  for (var i = 1; i < iterations; i++) {
    u = _hmacSha512(password, u);
    for (var j = 0; j < out.length; j++) {
      out[j] ^= u[j];
    }
  }
  return out;
}

Uint8List _hmacSha512(List<int> key, List<int> data) {
  return Uint8List.fromList(Hmac(sha512, key).convert(data).bytes);
}

Uint8List _u32(int value) {
  final data = ByteData(4)..setUint32(0, value, Endian.little);
  return data.buffer.asUint8List();
}

File: packaging/2026.07.03/drivers/elixir/runtime/elixir/stage/test/auth_bootstrap_contract_test.exs
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

defmodule ScratchBirdAuthBootstrapContractTest do
  use ExUnit.Case

  alias ScratchBird.{Connection, Protocol}

  @manager_protocol_magic 0x42444253
  @manager_protocol_version 0x0101
  @manager_header_size 12
  @mcp_msg_status_response 0x64
  @mcp_msg_hello 0x65

  test "probe_auth_surface reports direct SCRAM_SHA_512" do
    server =
      start_loopback_server(fn socket ->
        {:ok, msg} = recv_protocol_message(socket)
        assert msg.type == Protocol.message_type(:startup)

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_request),
          auth_request_payload(4)
        )
      end)

    {:ok, probe} =
      ScratchBird.probe_auth_surface(
        host: "127.0.0.1",
        port: server.port,
        database: "db",
        user: "user",
        sslmode: "disable"
      )

    assert probe.reachable
    assert probe.ingress_mode == "direct"
    assert probe.required_method == "SCRAM_SHA_512"
    assert probe.required_plugin_method_id == "scratchbird.auth.scram_sha_512"
    assert length(probe.admitted_methods) == 1
    [method] = probe.admitted_methods
    assert method.wire_method == "SCRAM_SHA_512"
    assert method.executable_locally
    assert probe.additional_continuation_possible
    await_loopback_server(server)
  end

  test "probe_auth_surface reports manager TOKEN ingress" do
    server =
      start_loopback_server(fn socket ->
        {:ok, msg_type, _payload} = recv_manager_frame(socket)
        assert msg_type == @mcp_msg_hello
        send_manager_frame(socket, @mcp_msg_status_response, <<>>)
      end)

    {:ok, probe} =
      ScratchBird.probe_auth_surface(
        host: "127.0.0.1",
        port: server.port,
        front_door_mode: "manager_proxy",
        database: "db",
        user: "admin",
        sslmode: "disable"
      )

    assert probe.reachable
    assert probe.ingress_mode == "manager_proxy"
    assert probe.required_method == "TOKEN"
    assert probe.required_plugin_method_id == "scratchbird.auth.authkey_token"
    assert length(probe.admitted_methods) == 1
    [method] = probe.admitted_methods
    assert method.wire_method == "TOKEN"
    assert method.executable_locally
    assert probe.additional_continuation_possible
    await_loopback_server(server)
  end

  test "connect tracks resolved SCRAM_SHA_512 auth context" do
    server =
      start_loopback_server(fn socket ->
        {:ok, msg} = recv_protocol_message(socket)
        assert msg.type == Protocol.message_type(:startup)

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_request),
          auth_request_payload(4)
        )

        {:ok, first_response} = recv_protocol_message(socket)
        assert first_response.type == Protocol.message_type(:auth_response)
        client_first = first_response.payload |> IO.iodata_to_binary() |> to_string()
        client_nonce = extract_scram_nonce(client_first)
        server_first = "r=#{client_nonce}server,s=#{Base.encode64("elixir-salt")},i=4096"

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_continue),
          auth_continue_payload(4, 0, server_first)
        )

        {:ok, final_response} = recv_protocol_message(socket)
        assert final_response.type == Protocol.message_type(:auth_response)
        assert byte_size(final_response.payload) > 0

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_ok),
          auth_ok_payload(""),
          session_id()
        )

        send_protocol_message(
          socket,
          Protocol.message_type(:ready),
          ready_payload(),
          session_id()
        )
      end)

    {:ok, conn} =
      Connection.connect(
        host: "127.0.0.1",
        port: server.port,
        database: "db",
        user: "user",
        password: "secret",
        sslmode: "disable"
      )

    context = ScratchBird.get_resolved_auth_context(conn)
    assert context.ingress_mode == "direct"
    assert context.resolved_auth_method == "SCRAM_SHA_512"
    assert context.resolved_auth_plugin_id == "scratchbird.auth.scram_sha_512"
    assert context.manager_authenticated == false
    assert context.attached == true

    Connection.close(conn)
    await_loopback_server(server)
  end

  test "connect tracks resolved TOKEN auth context" do
    server =
      start_loopback_server(fn socket ->
        {:ok, msg} = recv_protocol_message(socket)
        assert msg.type == Protocol.message_type(:startup)

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_request),
          auth_request_payload(5)
        )

        {:ok, auth_response} = recv_protocol_message(socket)
        assert auth_response.type == Protocol.message_type(:auth_response)
        assert auth_response.payload == "token-123"

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_ok),
          auth_ok_payload(""),
          session_id()
        )

        send_protocol_message(
          socket,
          Protocol.message_type(:ready),
          ready_payload(),
          session_id()
        )
      end)

    {:ok, conn} =
      Connection.connect(
        host: "127.0.0.1",
        port: server.port,
        database: "db",
        user: "user",
        sslmode: "disable",
        auth_token: "token-123"
      )

    context = ScratchBird.get_resolved_auth_context(conn)
    assert context.resolved_auth_method == "TOKEN"
    assert context.resolved_auth_plugin_id == "scratchbird.auth.authkey_token"
    assert context.attached == true

    Connection.close(conn)
    await_loopback_server(server)
  end

  test "connect fails closed for PEER auth" do
    server =
      start_loopback_server(fn socket ->
        {:ok, msg} = recv_protocol_message(socket)
        assert msg.type == Protocol.message_type(:startup)

        send_protocol_message(
          socket,
          Protocol.message_type(:auth_request),
          auth_request_payload(6)
        )
      end)

    assert_raise RuntimeError, ~r/requires external broker support/, fn ->
      Connection.connect(
        host: "127.0.0.1",
        port: server.port,
        database: "db",
        user: "user",
        sslmode: "disable"
      )
    end

    await_loopback_server(server)
  end

  defp start_loopback_server(handler) do
    {:ok, listener} =
      :gen_tcp.listen(0, [
        :binary,
        packet: :raw,
        active: false,
        reuseaddr: true,
        ip: {127, 0, 0, 1}
      ])

    {:ok, port} = :inet.port(listener)

    task =
      Task.async(fn ->
        {:ok, socket} = :gen_tcp.accept(listener)

        try do
          handler.(socket)
          :gen_tcp.close(socket)
          :gen_tcp.close(listener)
          :ok
        rescue
          error ->
            :gen_tcp.close(socket)
            :gen_tcp.close(listener)
            reraise(error, __STACKTRACE__)
        catch
          kind, reason ->
            :gen_tcp.close(socket)
            :gen_tcp.close(listener)
            raise "loopback server caught #{inspect(kind)}: #{inspect(reason)}"
        end
      end)

    %{port: port, task: task}
  end

  defp await_loopback_server(server) do
    Task.await(server.task, 1_000)
  end

  defp recv_protocol_message(socket) do
    with {:ok, header_bin} <- :gen_tcp.recv(socket, Protocol.header_size(), 1_000),
         {:ok, header} <- Protocol.decode_header(header_bin),
         {:ok, payload} <- :gen_tcp.recv(socket, header.length, 1_000) do
      {:ok, Map.put(header, :payload, payload)}
    end
  end

  defp send_protocol_message(socket, type, payload, attachment_id \\ <<0::128>>) do
    encoded =
      Protocol.encode_message(
        %{
          type: type,
          flags: 0,
          length: byte_size(payload),
          sequence: 0,
          attachment_id: attachment_id,
          txn_id: 0
        },
        payload
      )

    :ok = :gen_tcp.send(socket, encoded)
  end

  defp recv_manager_frame(socket) do
    with {:ok, header} <- :gen_tcp.recv(socket, @manager_header_size, 1_000),
         {:ok, msg_type, payload_len} <- parse_manager_header(header),
         {:ok, payload} <- :gen_tcp.recv(socket, payload_len, 1_000) do
      {:ok, msg_type, payload}
    end
  end

  defp send_manager_frame(socket, msg_type, payload) do
    frame =
      <<@manager_protocol_magic::little-32, @manager_protocol_version::little-16, msg_type::8,
        0::8, byte_size(payload)::little-32, payload::binary>>

    :ok = :gen_tcp.send(socket, frame)
  end

  defp parse_manager_header(
         <<@manager_protocol_magic::little-32, @manager_protocol_version::little-16, msg_type::8,
           _::8, payload_len::little-32>>
       ) do
    {:ok, msg_type, payload_len}
  end

  defp parse_manager_header(_), do: {:error, :invalid_manager_header}

  defp auth_request_payload(method), do: <<method::8, 0::24>>

  defp auth_continue_payload(method, stage, value) do
    data = IO.iodata_to_binary(value)
    <<method::8, stage::8, 0::16, byte_size(data)::little-32, data::binary>>
  end

  defp auth_ok_payload(info) do
    info = IO.iodata_to_binary(info)
    <<0::128, byte_size(info)::little-32, info::binary>>
  end

  defp ready_payload, do: <<0::8, 0::24, 0::little-64>>

  defp session_id,
    do:
      <<0x10, 0x32, 0x54, 0x76, 0x98, 0xBA, 0xDC, 0xFE, 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD,
        0xEF>>

  defp extract_scram_nonce("n,," <> bare) do
    bare
    |> String.split(",", trim: true)
    |> Enum.find_value(fn part ->
      case String.split(part, "=", parts: 2) do
        ["r", value] -> value
        _ -> nil
      end
    end)
  end

  defp extract_scram_nonce(value) do
    flunk("unable to extract SCRAM nonce from #{inspect(value)}")
  end
end
